src/cargo/ops/cargo_new.rs

```diff
@@ -412,74 +510,84 @@ fn mk(config: &Config, opts: &MkOptions) -> CargoResult<()> {
         (None, None, name, None) => name,
     };
 
-    let mut cargotoml_path_specifier = String::new();
-
-    // Calculare what [lib] and [[bin]]s do we need to append to Cargo.toml
+    let template_dir = try!(get_input_template(config, opts));
+    let template_path = match opts.template {
+        Some(template) => template_dir.path().join(template),
+        None => template_dir.path().to_path_buf()
+    };
 
-    for i in &opts.source_files {
-        if i.bin {
-            if i.relative_path != "src/main.rs" {
-                cargotoml_path_specifier.push_str(&format!(r#"
-[[bin]]
-name = "{}"
-path = {}
-"#, i.target_name, toml::Value::String(i.relative_path.clone())));
-            }
-        } else {
-            if i.relative_path != "src/lib.rs" {
-                cargotoml_path_specifier.push_str(&format!(r#"
-[lib]
-name = "{}"
-path = {}
-"#, i.target_name, toml::Value::String(i.relative_path.clone())));
-            }
-        }
+    // make sure that the template exists
+    if fs::metadata(&template_path).is_err() {
+        return Err(human(format!("Template `{}` not found", template_path.display())))
```

Comment:
You can use the `bail!` macro here for a bit more ergonomic bailing out, and also stylistically Cargo error messages start with a lowercase letter.